stdlib/public/core/AutoDiff.swift

```diff
@@ -495,6 +495,246 @@ public func gradient<T, U, V, R>(
+
+  /// The underlying base value, type-erased to `Any`.
+  var _typeErasedBase: Any { get }
+
+  /// Returns the underlying value unboxed to the given type, if possible.
+  func _unboxed<U>(to type: U.Type) -> U?
+    where U : Differentiable, U.TangentVector == U,
+          U.AllDifferentiableVariables == U,
+          // NOTE: The requirement below should be defined on `Differentiable`.
+          // But it causes a crash due to generic signature minimization bug.
+          U.CotangentVector == U.CotangentVector.AllDifferentiableVariables
+}
+
+extension _AnyDerivativeBox {
+  /// Returns true if the underlying value has type `AnyDerivative.OpaqueZero`.
+  func _isOpaqueZero() -> Bool {
+    return _unboxed(to: AnyDerivative.OpaqueZero.self) != nil
+  }
+}
+
+@inline(never)
+@usableFromInline
+internal func _derivativeTypeMismatch(
+  _ x: Any.Type, _ y: Any.Type, file: StaticString = #file, line: UInt = #line
+) -> Never {
+  fatalError("Derivative type mismatch: \(x) and \(y)", file: file, line: line)
+}
+
+internal struct _ConcreteDerivativeBox<T> : _AnyDerivativeBox
+  where T : Differentiable, T.TangentVector == T,
+        T.AllDifferentiableVariables == T,
+        // NOTE: The requirement below should be defined on `Differentiable`.
+        // But it causes a crash due to generic signature minimization bug.
+        T.CotangentVector == T.CotangentVector.AllDifferentiableVariables
+{
+  /// The underlying base value.
+  var _base: T
+
+  init(_ base: T) {
+    self._base = base
+  }
+
+  /// The underlying base value, type-erased to `Any`.
+  var _typeErasedBase: Any {
+    return _base
+  }
+
+  func _unboxed<U>(to type: U.Type) -> U?
+    where U : Differentiable, U.TangentVector == U,
+          U.AllDifferentiableVariables == U,
+          // NOTE: The requirement below should be defined on `Differentiable`.
+          // But it causes a crash due to generic signature minimization bug.
+          U.CotangentVector == U.CotangentVector.AllDifferentiableVariables
+  {
+    return (self as? _ConcreteDerivativeBox<U>)?._base
+  }
+
+  // `Equatable` requirements (implied by `AdditiveArithmetic`).
+
+  func _isEqual(to other: _AnyDerivativeBox) -> Bool {
+    return _base == other._unboxed(to: T.self)
+  }
+
+  func _isNotEqual(to other: _AnyDerivativeBox) -> Bool {
+    return _base != other._unboxed(to: T.self)
+  }
+
+  // `AdditiveArithmetic` requirements.
+
+  static var _zero: _AnyDerivativeBox {
+    return _ConcreteDerivativeBox(T.zero)
+  }
+
+  static var _dualSpaceZero: _AnyDerivativeBox {
+    return _ConcreteDerivativeBox<T.CotangentVector>(T.CotangentVector.zero)
+  }
+
+  func _adding(_ x: _AnyDerivativeBox) -> _AnyDerivativeBox {
+    // 0 + x = x
+    if _isOpaqueZero() {
+      return x
+    }
+    // y + 0 = y
+    if x._isOpaqueZero() {
+      return self
+    }
+    guard let xBase = x._unboxed(to: T.self) else {
+      _derivativeTypeMismatch(type(of: self), type(of: x))
```

Comment:
`type(of: self)` or `type(of: x)` are types of the concrete box, not the user type. We should show the user type instead, i.e. `T.self` and `type(of: x._typeErasedBase)` respectively.

Same for other occurrences of `_derivativeTypeMismatch` calls. I'd suggest trying those out to verify the any internal types won't be shown.

Reply:
Good point - changed to `T.self` and `type(of: x._typeErasedBase)` everywhere.
Unfortunately, fully-qualified types are not printed:
```
Fatal error: Derivative type mismatch: TangentVector and TangentVector
```